pkg/provisioners/kafka/controller/channel/reconcile.go

```diff
@@ -0,0 +1,141 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package channel
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Channel resource
+// with the current status of the resource.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.log.Info("Reconciling channel", "request", request)
+	channel := &v1alpha1.Channel{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, channel)
+
+	if errors.IsNotFound(err) {
+		r.log.Info("could not find channel", "request", request)
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		r.log.Error(err, "could not fetch Channel", "request", request)
+		return reconcile.Result{}, err
+	}
+
+	original := channel.DeepCopy()
+
+	// Reconcile this copy of the Channel and then write back any status
+	// updates regardless of whether the reconcile error out.
+	err = r.reconcile(channel)
+	if !equality.Semantic.DeepEqual(original.Status, channel.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+		if _, err := r.updateStatus(channel); err != nil {
+			r.log.Info("Failed to update channel status", "error", err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Requeue if the resource is not ready:
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(channel *v1alpha1.Channel) error {
+	// See if the channel has been deleted
+	accessor, err := meta.Accessor(channel)
+	if err != nil {
+		r.log.Info("failed to get metadata", "error", err)
+		return err
+	}
+	deletionTimestamp := accessor.GetDeletionTimestamp()
+	if deletionTimestamp != nil {
+		r.log.Info(fmt.Sprintf("DeletionTimestamp: %v", deletionTimestamp))
+		//TODO: Handle deletion
+		return nil
+	}
+
+	// Skip Channel as it is not targeting any provisioner
+	if channel.Spec.Provisioner == nil || channel.Spec.Provisioner.Ref == nil {
+		return nil
+	}
+
+	// Skip channel not managed by this provisioner
+	provisionerRef := channel.Spec.Provisioner.Ref
+	clusterProvisioner, err := r.getClusterProvisioner()
```

Comment:
this is a cool idea! I wonder if the flexibility of setting the name of the provisioner should be a argument though, what does it mean for the controller to change which provisioner it is servicing while it is running?

Reply:
I removed the flexibility of the setting the provisioner name with the latest commit. I don't find a use case as of now. And this will be in line with the in-memory channel provisioner.